# Project Spam Classifier

### Data_set has been Downloaded from UCI Machine Learning Repositroy and name of data set is SMS Spam Collection Data Set

In [1]:
import pandas as pd
messages=pd.read_csv('SMSSpamCollection',sep='\t',names=["label","message"])
# Giving names to the two coloumns of the data

In [2]:
messages[:3]

label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...

### ham means message is not spam  Moreover message and ham are tab seprated in the dataset. Hence we read our dataset by using '\t' and then name them

In [3]:
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer


### Columns are actually tab seprated , however words in 2nd coloumns are space seprated 

In [4]:
# Cleaning the data

ps = PorterStemmer()
corpus = []
for i in range(0, len(messages)):
    review = re.sub('[^a-zA-Z]', ' ', messages['message'][i])
    
    # So that same word having lower and capital alphabets is not considered to be different
    review = review.lower()
    
    # To have the review as the list of words of that particular sentence
    review = review.split()
    
    # Stop words (i.e in,a,at,they) are of no use
    review = [ps.stem(word) for word in review if not word in stopwords.words('english')]
    review = ' '.join(review)
    corpus.append(review)
    

### Stemming is the technique to have the base root form of the word

### Apply stemming first as it take less amount of time but if accuracy is bad then reconsider it and apply lemmitization

## Creating bag of words

In [7]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer()
X = cv.fit_transform(corpus).toarray()
print(X.shape)


# We are using max_features as we only need frequently used words not all the words are of use
# As some of the words may be used only once or twice but still they can increase the computational time

cv = CountVectorizer(max_features=2500)
X = cv.fit_transform(corpus).toarray()

print(X.shape)


(5572, 6296)
(5572, 2500)


## i.e we have Reduce the word count to 2500 most used words from 6296

In [14]:
# We are converting "ham" and "spam" to dummy values i.e 1 and 0 by using inbuilt library of pandas i.e get_dummies
y=pd.get_dummies(messages['label'])


print(y[:5])
print('\n')

# Converting to one coloumn from two coloumns as both are actually same
y=y.iloc[:,1].values
print(y[:5])


   ham  spam
0    1     0
1    1     0
2    0     1
3    1     0
4    1     0


[0 0 1 0 0]


## Training the split data

In [15]:
# Train Test Split by using the inbuilt libraries of skiket learn i.e train_test_split library

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)


## Training the model using Naive Bayes Classifier

In [16]:
# Training the Model using using inbuilt library

from sklearn.naive_bayes import MultinomialNB
spam_detect_model = MultinomialNB().fit(X_train, y_train)


In [18]:
y_pred=spam_detect_model.predict(X_test)
print(y_pred[:5])
print(y_test[:5])

[0 1 0 0 0]
[0 1 0 0 0]


### We can see that we are getting the same values for pred and test data i.e our classifier is working

## Checking the correct Predection of the model using confusion Matrix

In [20]:
from sklearn.metrics import confusion_matrix
accuracy=confusion_matrix(y_test,y_pred)

print(accuracy)

[[946   9]
 [  7 153]]


### We have 946 ( 0 out of 0 ) and 153 (1 out of 1) i.e total correct predection is 946+153

In [21]:
from sklearn.metrics import accuracy_score
a_score=accuracy_score(y_test,y_pred)
print(a_score)

0.9856502242152466


## 98% is a good score hence stemming is working no need to do lemmitization as lemmitization increases computational time